### ISSUE TITLE
Dialyzer crashing due to infinite loop being detected during analysis

### ISSUE THREAD

**Issue by TD5** [erlang/otp#5410]
**Describe the bug**
Given specific input code for analysis, Dialyzer will crash after it detects it is stuck in an infinite loop. The bug seems to require a recursive lambda being used in a certain context to be triggered.

**To Reproduce**
Have Dialyzer analyse this snippet:

```
-spec bar(fun((pos_integer()) -> term())) -> term().
bar(Fun) ->
  Fun.

foo_bad() ->
    bar(fun F(_) -> F(1) end).
```

It will crash with:
> "Infinite loop detected, please report this bug."

**Expected behavior**
Ideally, Dialyzer wouldn't crash. Either it would locate a specific error, or accept the code.

**Affected versions**
Recent commits to erlang/otp. At least commit 8076ba28b8b12d495049c5b6145cabb308472367.

**Additional context**
I can see from recent commits to Dialyzer that [this is somewhat intended](https://github.com/erlang/otp/commit/8076ba28b8b12d495049c5b6145cabb308472367#diff-72a49edc258056a5277e70d01e2c8ff44dbe388efa722a5752df52cce8694e46) (insofar as to highlight the bug which was previously potentially unseen). Unfortunately, this makes it very hard to analyse a big codebase which uses Erlang's features in lots of obscure but presumably valid ways.

There is a work-around: Factoring out the recursive lambda into a named top-level function. i.e. the following does not cause Dialyzer to crash:

```
-spec bar(fun((pos_integer()) -> term())) -> term().
bar(Fun) ->
  Fun.

f(_) -> f(1).

foo_ok() ->
    bar(fun f/1).
```

**Comment by jhogberg** [comment]
Thanks for your report, I've fixed it in #5475. Feel free to try it out.

**Comment by TD5** [comment]
Thanks for the fix!
